hail/src/main/scala/is/hail/expr/types/physical/PDict.scala

```diff
@@ -4,33 +4,20 @@ import is.hail.annotations._
 import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TDict
 
-final case class PDict(keyType: PType, valueType: PType, required: Boolean = false)
-  extends PArrayBackedContainer(PCanonicalArray(PStruct(required = true, "key" -> keyType, "value" -> valueType), required)) {
-  val elementType = arrayRep.elementType.asInstanceOf[PStruct]
+object PDict {
+  def apply(keyType: PType, valueType: PType, required: Boolean = false) = PCanonicalDict(keyType, valueType, required)
+}
 
+abstract class PDict extends PContainer {
   lazy val virtualType: TDict = TDict(keyType.virtualType, valueType.virtualType, required)
 
-  def _asIdent = s"dict_of_${keyType.asIdent}AND${valueType.asIdent}"
-  def _toPretty = s"Dict[$keyType, $valueType]"
+  val keyType: PType
+  val valueType: PType
 
-  override def pyString(sb: StringBuilder): Unit = {
-    sb.append("dict<")
-    keyType.pyString(sb)
-    sb.append(", ")
-    valueType.pyString(sb)
-    sb.append('>')
-  }
+  def copy(keyType: PType = this.keyType, valueType: PType = this.valueType, required: Boolean = this.required): PDict
 
-  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean = false) {
-    sb.append("Dict[")
-    keyType.pretty(sb, indent, compact)
-    if (compact)
-      sb += ','
-    else
-      sb.append(", ")
-    valueType.pretty(sb, indent, compact)
-    sb.append("]")
-  }
+  def elementType: PStruct = ???
```

Comment:
this is bad practice. Leave abstract instead of `= ???`.

Reply:
I was not able to find a better away. Abstract doesn't work, Scala will not upcast from PType (aka make this a Trait and declare  `abstract elementType: PStruct`. Similarly, setting a overridable `type ElementType: PType ; def elementType: PType` in PType and overriding ElementType in PDict did not pass type check. In order to have this be PStruct, as far as I can tell, we either need to do this, or cast ever instance of elementType used by a caller to PStruct.

Reply:
<img width="626" alt="Screen Shot 2019-12-15 at 4 17 23 PM" src="https://user-images.githubusercontent.com/5543229/70869191-63c93280-1f56-11ea-95c6-0d4afa2164fc.png">

Reply:
what is `abstract elementType: PStruct`? I've never seen that pattern before.

Does 
```
def elementType: PBaseStruct
```
not work?

Reply:
Just issued a PR for this. I thought I had checked this case, but re-checking, type checked fine. Thanks for the comment.